lib/THNN/generic/TemporalRowConvolutionMM.c

```diff
@@ -0,0 +1,488 @@
+	for (c = 0; c < inputFrameSize; ++c) {
+		size_t kw, x;
+		long long ix = 0;
+
+		for (kw = 0; kw < kW; ++kw) {
+			real *src = finput_data
+			            + c * (kW * nOutputFrame)
+			            + kw * (nOutputFrame);
+			real *dst = input_data + c * (nInputFrame);
+
+			ix = (long long)(kw);
+			if (dW == 1) {
+				THVector_(add)(dst + (size_t)(ix), src, 1, nOutputFrame);
+			} else {
+				for (x = 0; x < nOutputFrame; ++x) {
+					THVector_(add)(dst + (size_t)(ix + x * dW),
+					               src + (size_t)(x), 1, 1);
+				}
+			}
+		}
+	}
+}
+
+static void THNN_(unfolded_copy_row)(
+	THTensor *finput,
+	THTensor *input,
+	int kW,
+	int dW,
+	int padW,
+	long inputFrameSize,
+	long nInputFrame,
+	long nOutputFrame) {
+
+	long k;
+	real *input_data = THTensor_(data)(input);
+	real *finput_data = THTensor_(data)(finput);
+
+#pragma omp parallel for private(k)
+	for (k = 0; k < inputFrameSize * kW; ++k) {
+		size_t c = k / kW;
+		size_t rest = k % kW;
+		size_t kw = rest % kW;
+		size_t x;
+		long long ix;
+		real *dst = finput_data + c * (kW * nOutputFrame) + kw * (nOutputFrame);
+		real *src = input_data + c * (nInputFrame);
+
+		ix = (long long)(kw);
+		if (dW == 1) {
+			memcpy(dst, src+(size_t)(ix), sizeof(real) * (nOutputFrame));
+		} else {
+			for (x = 0; x < nOutputFrame; ++x) {
+				memcpy(dst + (size_t)(x), src + (size_t)(ix + x * dW),
+				       sizeof(real) * 1);
+			}
+		}
+	}
+}
+
+static void THNN_(TemporalRowConvolutionMM_updateOutput_frame)(
+	THTensor *input,
+	THTensor *output,
+	THTensor *weight,
+	THTensor *bias,
+	THTensor *finput,
+	int kW,
+	int dW,
+	int padW,
+	long inputFrameSize,
+	long nInputFrame,
+	long nOutputFrame) {
+
+	long i;
+	THTensor *output3d;
+
+	THNN_(unfolded_copy_row)(finput, input, kW, dW, padW,
+	                         inputFrameSize, nInputFrame, nOutputFrame);
+
+	output3d = THTensor_(newWithStorage3d)(
+		output->storage, output->storageOffset,
+		inputFrameSize, -1,
+		1, -1,
+		nOutputFrame, -1);
+
+	if (bias != NULL) {
+		for (i = 0; i < inputFrameSize; ++i)
+			THVector_(fill)
+			        (output->storage->data + output->storageOffset
+			        + output->stride[0] * i,
+			        THTensor_(get1d)(bias, i), nOutputFrame);
+	} else {
+		THTensor_(zero)(output);
+	}
+
+	THTensor_(baddbmm)(output3d, 1, output3d, 1, weight, finput);
+
+	THTensor_(free)(output3d);
+}
+
+
+void THNN_(TemporalRowConvolutionMM_updateOutput)(
+	THNNState *state,
+	THTensor *input,
+	THTensor *output,
+	THTensor *weight,
+	THTensor *bias,
+	THTensor *finput,
+	THTensor *fgradInput,   // unused here but needed for Cuda
+	int kW,
+	int dW,
+	int padW) {
+
+	THNN_(TemporalRowConvolutionMM_shapeCheck)(
+		state, input, NULL, weight, bias, kW, dW, padW);
+
+	int freeWeight = THNN_(view_weight_rowconv)(&weight);
+
+	int ndim = input->nDimension;
+	int dimS = 0;
+	int dimF = 1;
+
+	if (ndim == 3) {
+		++dimS;
+		++dimF;
+	}
+
+	THTensor *tinput = THTensor_(newTranspose)(input, dimS, dimF);
```

Comment:
the way you wrote this function with all the transposes, it'd be a shame to not also have the function support feature-dimension be 2nd dim optionally (the way it's in pytorch, i.e. Batch x Channels x Width), because when you want to use this in to pytorch, Conv1d's output can be sent directly into this.
Consider adding a boolean flag to the function signature that controls this transposition